File: src/c/pointer.c
// 8️⃣ C Pointer Types and Manual Memory Management
// 
// Compile: gcc -O2 -S src/c/pointer.c -o target/pointer.s
// Assembly file: target/pointer.s
//
// Key assembly differences to observe:
// - No compile-time borrow checking, all pointer access is manual
// - const pointers are compiler hints, not enforced at runtime
// - Manual malloc/free required for heap allocation
// - Multiple pointers to same memory have no automatic reference counting
// - Volatile pointers prevent compiler optimizations
// - No automatic cleanup - programmer responsible for memory management
// - Risk of dangling pointers, use-after-free, and memory leaks

#include <stdio.h>
#include <stdlib.h>

// 1. Const pointer - similar to immutable reference but not enforced
int const_pointer_example() {
    int value = 42;
    const int *ptr = &value;  // const pointer, compiler hint only
    return *ptr;  // Direct memory access, no runtime checking
}

// 2. Mutable pointer - similar to mutable reference but no exclusivity
int mutable_pointer_example() {
    int value = 42;
    int *ptr = &value;  // Mutable pointer, no exclusivity guarantees
    *ptr += 10;         // Direct mutation, no borrow checking
    return *ptr;
}

// 3. Heap allocation - similar to Box<T> but manual cleanup required
int heap_pointer_example() {
    int *heap_ptr = malloc(sizeof(int));  // Manual heap allocation
    if (heap_ptr == NULL) return -1;      // Manual error checking required
    
    *heap_ptr = 42;
    int result = *heap_ptr + 10;
    
    free(heap_ptr);  // Manual deallocation required to prevent leaks
    return result;
}

// 4. Multiple pointers to same memory - simulates Rc<T> behavior
int shared_pointer_example() {
    int *original = malloc(sizeof(int));
    if (original == NULL) return -1;
    
    *original = 42;
    int *alias1 = original;  // Multiple pointers to same memory
    int *alias2 = original;  // No automatic reference counting
    
    int result = *alias1 + *alias2;  // Both access same memory location
    
    free(original);  // Only free once, programmer must track ownership
    // alias1 and alias2 are now dangling pointers (unsafe to use)
    return result;
}

// 5. Pointer arithmetic - array-style access
int pointer_arithmetic_example() {
    int array[3] = {10, 20, 30};
    int *ptr = array;  // Points to first element
    
    int result = 0;
    result += *ptr;        // array[0]
    result += *(ptr + 1);  // array[1] via pointer arithmetic
    result += *(ptr + 2);  // array[2] via pointer arithmetic
    
    return result;  // No bounds checking, potential buffer overflow
}

// 6. Volatile pointer - prevents compiler optimization, similar to raw pointers
int volatile_pointer_example() {
    int value = 42;
    volatile int *vol_ptr = &value;  // Volatile prevents optimization
    
    int mutable_value = 100;
    volatile int *mut_vol_ptr = &mutable_value;
    
    int read_val = *vol_ptr;      // Volatile read
    *mut_vol_ptr += 10;           // Volatile write
    
    return read_val + *mut_vol_ptr;
}

// 7. Function pointer - demonstrates pointer to code
int add_ten(int x) {
    return x + 10;
}

int function_pointer_example() {
    int (*func_ptr)(int) = add_ten;  // Pointer to function
    return func_ptr(42);  // Call through function pointer
}

int main() {
    // Call all functions to prevent compiler optimization
    printf("Const pointer: %d\n", const_pointer_example());
    printf("Mutable pointer: %d\n", mutable_pointer_example());
    printf("Heap pointer: %d\n", heap_pointer_example());
    printf("Shared pointer: %d\n", shared_pointer_example());
    printf("Pointer arithmetic: %d\n", pointer_arithmetic_example());
    printf("Volatile pointer: %d\n", volatile_pointer_example());
    printf("Function pointer: %d\n", function_pointer_example());
    
    return 0;
} 

File: src/rust/pointer.rs
// 8️⃣ Rust Pointer Types and Memory Management
// 
// Compile: cargo rustc --bin pointer --release -- --emit=asm
// Assembly file: target/release/deps/pointer-*.s
//
// Key assembly differences to observe:
// - Immutable/mutable references have zero runtime cost
// - Box<T> generates heap allocation and automatic deallocation
// - Rc<T> includes reference counting with atomic operations
// - Arc<T> uses atomic reference counting for thread safety
// - Raw pointers require unsafe blocks and generate direct memory access
// - Pin<Box<T>> prevents movement but has same allocation cost as Box<T>
// - Rust's ownership system prevents use-after-free at compile time

use std::rc::Rc;
use std::sync::Arc;
use std::pin::Pin;

// 1. Immutable reference - zero cost abstraction
fn immutable_reference_example() -> i32 {
    let value = 42;
    let reference: &i32 = &value;  // Immutable borrow, compile-time checked
    *reference  // Dereference has no runtime cost
}

// 2. Mutable reference - exclusive access guaranteed
fn mutable_reference_example() -> i32 {
    let mut value = 42;
    let reference: &mut i32 = &mut value;  // Exclusive mutable borrow
    *reference += 10;  // Safe mutation through reference
    *reference
}

// 3. Box<T> - heap allocation with automatic cleanup
fn box_pointer_example() -> i32 {
    let boxed_value: Box<i32> = Box::new(42);  // Heap allocation
    let result = *boxed_value + 10;
    result  // Box automatically deallocated when dropped
}

// 4. Rc<T> - reference counted smart pointer (single-threaded)
fn rc_pointer_example() -> i32 {
    let rc_value: Rc<i32> = Rc::new(42);  // Reference count = 1
    let rc_clone = Rc::clone(&rc_value);   // Reference count = 2
    let result = *rc_value + *rc_clone;    // Both point to same data
    result  // Reference count decremented automatically
}

// 5. Arc<T> - atomic reference counted smart pointer (thread-safe)
fn arc_pointer_example() -> i32 {
    let arc_value: Arc<i32> = Arc::new(42);  // Atomic reference count = 1
    let arc_clone = Arc::clone(&arc_value);  // Atomic increment to 2
    let result = *arc_value + *arc_clone;    // Thread-safe shared ownership
    result  // Atomic decrement when dropped
}

// 6. Raw pointers - unsafe but direct memory access
fn raw_pointer_example() -> i32 {
    let value = 42;
    let const_ptr: *const i32 = &value;  // Raw const pointer
    let mut mutable_value = 100;
    let mut_ptr: *mut i32 = &mut mutable_value;  // Raw mutable pointer
    
    unsafe {
        // Direct memory access without safety checks
        let read_const = *const_ptr;      // Read through const pointer
        *mut_ptr += 10;                   // Write through mutable pointer
        read_const + *mut_ptr
    }
}

// 7. Pin<Box<T>> - prevents movement, useful for self-referential structs
fn pinned_pointer_example() -> i32 {
    let boxed_value = Box::new(42);
    let pinned: Pin<Box<i32>> = Pin::new(boxed_value);  // Cannot be moved
    *pinned  // Dereference: Pin<Box<T>> auto-derefs to T
}

fn main() {
    // Call all functions to prevent compiler optimization
    println!("Immutable reference: {}", immutable_reference_example());
    println!("Mutable reference: {}", mutable_reference_example());
    println!("Box pointer: {}", box_pointer_example());
    println!("Rc pointer: {}", rc_pointer_example());
    println!("Arc pointer: {}", arc_pointer_example());
    println!("Raw pointer: {}", raw_pointer_example());
    println!("Pinned pointer: {}", pinned_pointer_example());
} 